projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+    std::ofstream coordinates;
+
+    //write headers  for csv
+    std::string csvpath = folderName + "/coordinates.csv";
+    coordinates.open (csvpath, std::ofstream::trunc);
+    coordinates  << "gps quality" << "," << "lat" << "," << "lon" << "," << "altitude"  << "," << "roll" << "," << "pitch" << "," << "yaw"  << ",";
+    coordinates << "image name" << "," << "timestamp" <<"\n";
+    coordinates.close();
+
+
+    BoBRobotics::BackgroundExceptionCatcher catcher;
+    catcher.trapSignals();
+
+    int i = 0;
+    BoBRobotics::Stopwatch sw, sw_timestamp;
+    std::array<degree_t, 3> angles;
+    // for x timestep
+    sw_timestamp.start();
+    for (;;) {
+
+        try {
+            catcher.check();
+            sw.start();
+            // get imu data
+            angles = imu.getEulerAngles();
+            float yaw = angles[0].value();
+            float pitch = angles[1].value();
+            float roll = angles[2].value();
+
+            // get gps data
+            BoBRobotics::GPS::GPSData data = gps.getGPSData();
+            BoBRobotics::MapCoordinate::GPSCoordinate coord = data.coordinate;
+            BoBRobotics::GPS::TimeStamp time = data.time;
+            int gpsQual = (int) data.gpsQuality; // gps quality
+
+            // output results
+    	   // if (i%5 == 0) { // only output once ever sec
+            std::cout << std::setprecision(10) << "GPS quality: " << gpsQual << " latitude: " << coord.lat.value()<< " longitude: " <<  coord.lon.value()  << " num sats: " << data.numberOfSatellites << " time: " << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value()/1000 << std::endl;
+
+	        std::ostringstream fileString, folderString;
+            fileString << folderName << "/image" << i << ".jpg";
+	        folderString << folderName << "/coordinates.csv";
+            std::string fileName = fileString.str();
+
+            //  headers:   gpsQuality|latitude|longitude|altitude|roll|pitch|yaw|imagename|timestamp
+            coordinates.open (folderString.str(), std::ofstream::app); // open coordinates.csv file and append
+            coordinates << std::setprecision(10) << std::fixed << gpsQual << "," << coord.lat.value() << "," << coord.lon.value() << "," << data.altitude.value()  << "," << roll << "," << pitch << "," << yaw  << ",";
+            coordinates << "image" << i << ".jpg" << "," << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value() <<"\n";
+            coordinates.close();
+
+            // poll from camera thread
+            cv::Mat originalImage;
+            mex.lock();
+            tmpImage.copyTo(originalImage);
+            mex.unlock();
+
+            if (originalImage.size().height > 0) {
+               // cv::resize(originalImage, originalImage, cv::Size(), 0.5, 0.5);
+               // cv::imshow("orig",resized);
+            }
+
+            std::thread imwriteThread{ &saveThread, std::move(originalImage), std::move(fileName) };
```

Comment:
I don't think this is a good idea - if you want to save every image just do it in the same thread as the camera (as I suggest above, I think it should be the main thread). If these worker threads are slower than this loop they'll be writing ``originalImage`` when it's in an ambiguous state as the next loop iteration will have started to mess with it. ``std::move`` is definitely not what you want here either